includes/sanitizers/class-amp-style-sanitizer.php

```diff
@@ -456,8 +456,29 @@ private function get_used_class_names() {
 			$classes .= ' ' . $class_attribute->nodeValue;
 		}
 
-		// Find all [class] attributes and capture the contents of any single- or double-quoted strings.
+		// Find all [class] attributes and capture the contents of any single- or double-quoted strings,
+		// as well as <amp-state> content that might be used.
 		foreach ( $this->xpath->query( '//*/@' . AMP_DOM_Utils::AMP_BIND_DATA_ATTR_PREFIX . 'class' ) as $bound_class_attribute ) {
+			if ( preg_match_all( '/([a-zA-Z0-9_]+?)\[\s*([\'"])([^\2]*?)\2\s*\]/', $bound_class_attribute->nodeValue, $matches ) ) {
+				$amp_states = $matches[1];
+				$properties = $matches[3];
+
+				foreach ( $amp_states as $index => $amp_state ) {
+					$state_element = $this->xpath->query( "//amp-state[@id = '{$amp_state}' ]/script" )->item( 0 );
+					if ( ! $state_element ) {
+						continue;
+					}
+
+					$state_data = json_decode( $state_element->textContent, true );
+
+					if ( isset( $state_data[ $properties[ $index ] ] ) ) {
+						$classes .= ' ' . implode( ' ', $state_data[ $properties[ $index ] ] );
+					}
+				}
+
+				continue;
+			}
+
```

Comment:
Could you provide an example for what this is solving for?

Reply:
I don't think it is solving anything right now.

Here's the current state of the code: I'm pretty much done with providing a generic replacement of the toggles and modals functionality of Twentytwenty. This means I'm not only using the markup of the theme to know where to use amp-state, amp-lightbox, etc, but it would also just work for people using twentytwenty as a base and extending it. If they add a new toggle or modal, it is automatically supported by AMP as well.

The one remaining issue is the styling right now, as I correctly add and remove the proper classes to the toggles, but these are still hidden because of a `display: none` style. I thought I could solve it by making sure the `active` class is not tree-shaken, but the behavior is more complex than that.

So the above code does not do much right now, and I will probably remove it again, because I need to solve this in a different way.

Reply:
Well, it does indeed solve the problem once the bugs are out... :)

Reply:
What are the classes that this is adding to `$classes`?

Perhaps this would be good to cherry-pick into a separate PR?

Reply:
Right now, this is not being used anymore. But it would be used again if we swicth back to using `<amp-bind>` for toggling classes, as we'll need it for making the `aria-expanded` toggling work (see https://github.com/ampproject/amp-wp/pull/3403#issuecomment-541666915).

In that case, this would be responsible for keeping the `active` class around, as it will not be visible in the initial markup but added via `setState()`.
I suggest removing this from here, and doing a separate a11y PR for `aria-expanded` once the a11y stuff for twentytwenty has settled.